packages/pageheader/components/PageHeader.tsx

```diff
@@ -0,0 +1,31 @@
+import * as React from "react";
+
+import { BreadcrumbProps } from "../../breadcrumb/components/Breadcrumb";
+
+export interface PageHeaderProps {
+  children: {
```

Comment:
Why did you choose to make `children` an object instead of just adding the props directly to the component?

Reply:
I have added it to children since I felt like the contents are children of the component not attributes which one would rather expose as an attribute property.

Reply:
I find it strange as well. This requires children to be an object, but the element itself is already a suitable object container. Why not just have title/breadcrumb/actions props? This limits the flexibility of the component in the future since children are usually understood to be react elements.

Reply:
Just so that we are all on the same page you would prefer to have something like that:
```JSX
<PageHeader breadcrumb={
  <Breadcrumb>
    <span>One</span>
    <span>Two</span>
  </Breadcrumb>
  } actions={[
  <Clickable key="Action1" action={action}>
    <div>Action 1</div>
  </Clickable>
]} />
```

I really feel the attribute props should not be used for ReactNodes but this is my opinion I am happy to discuss.

Reply:
That is it, yes. props is already a "named slot" I'm not sure what the benefit is in overtaking children for this purpose

Reply:
Yes this is correct. I feel like the children approach is better maintainable and overall a better user experience. Please compare the above code with the following:
```JSX
<PageHeader>
  {{
    breadcrumb: (
      <Breadcrumb>
        <span>One</span>
        <span>Two</span>
      </Breadcrumb>
    ),
    actions: [
      <Clickable key="Action1" action={action}>
        <div>Action 1</div>
      </Clickable>
    ]
  }}
</PageHeader>
```

To me I can better see where the PageHeader begins and ends. The named slots are similar to distinguish so there is no drawback nor a potential gain. Also semantically I feel like the breadcrumb and the actions are a children property of the PageHeader and not an attribute of the PageHeader. 

Here is a third proposal would you say that this one is also incorrect or would you say that it is favourable over the attribute approach?

```
<PageHeader>
  <Breadcrumb>
    <span>One</span>
    <span>Two</span>
  </Breadcrumb>
  <ActionList>
    <Clickable key="Action1" action={action}>
      <div>Action 1</div>
    </Clickable>
  </ActionList>
</PageHeader>
```

Reply:
I've heard some agreement that children "named slots" is a strange convention and I remain not convinced we should adopt it. You argued that you can better see where the PageHeader begins and ends, but this could also be taken to mean that an end tag is now necessary. Furthermore, there is a real drawback- you lose the flexibility of having normal, renderable children. I'm still fond of the props version, even as ReactNodes. Maybe we can get a fourth voice in here?

Reply:
I prefer the props pattern. IMO the children object pattern is unfamiliar and could be unnatural to others who are trying to consume this component.

Reply:
I tried to write up pros and cons, I figured that most of the pros are all applicable to putting the children into props as well. So it all comes down to wether or not we want to put children into the children prop or rather into other props. Although I want to raise that I think it is a huge anti pattern to put children into none children props. And if there are multiple `areas` within one component and one would not like to add the complexity of iterating through children to actually find the relevant areas. One might use this pattern to differentiate between different areas in a lot less complex manner. 
About the closing tag I don't feel that this is a drawback, since most HTML tags do need a closing tag this should be a widely used pattern. 
@natmegs I agree with your point that it is unfamiliar and that it feels unnatural if this would be the only component which is using this pattern which is correct right now. How would you feel about this pattern if this is a pattern we adopt for similar situations?

Reply:
I feel like props are already doing the same thing as the children object method, but without any additional boilerplate. My concern about it being unfamiliar/unnatural is not only just in comparison to other components in this library but more in terms of React patterns generally, so I don't think it would help to adopt this widely. I think it's commonly assumed that children will be a ReactNode, and for that reason it would be easy to misuse this component if it requires children that are an object of a particular shape.

Reply:
I thought a little more about that and discussed this with @DanielMSchmidt and @nLight today. I came to the conclusion that the object as a child pattern could also have performance issues as pointed out by @DanielMSchmidt, which is a good reason not to use it. But while talking about it we kind of came to the conclusion that passing ReactNode's as props is an anti pattern and most of the times when used we probably leave too much surface to the api. We probably want to pass something else for example for the `PageHeader` instead of expecting a `breadcrumb` and a list of `actions` we probably want to have a list of `breadcrumbElements` and a list of `actions`  which will change the type to look like this:
```TS
interface PageHeaderProps = {
  breadcrumbElements: React.ReactNode[];
  actions?: React.ReactNode[]; 
}
```

Also I agreed with @nLight that we should create guidelines on how to create components. Similar to `eslint` so that we have a team agreement on which patterns we use. This should unify on how components look. And this will erase eventually discussions about opinions. 

Just to mentioned it here at least we also came to the conclusion that if there is ever a need to pass ReactNode's as props one should use the composable pattern by exposing either pseudo components like `Column` or expect certain components e.g.:
```JS
<PageHeader>
  <Breadcrumb>
    <span>One</span>
    <span>Two</span>
  </Breadcrumb>
  <ActionList>
    <Clickable key="Action1" action={action}>
      <div>Action 1</div>
    </Clickable>
  </ActionList>
</PageHeader>
```

This will be reflected eventually in a guideline design doc.

Reply:
> Also I agreed with @nLight that we should create guidelines on how to create components.

Keep me posted - I'd love to be a part of creating this guideline design doc.
---

> we kind of came to the conclusion that passing ReactNode's as props is an anti pattern and most of the times when used we probably leave too much surface to the api

100% agreed that it creates a big surface area. My favorite [blog post about component APIs](https://jxnblk.com/blog/defining-component-apis-in-react/) touches on this topic.